File: test/hs/TestTuples.hs
module TestTuples where

import Foreign.C.Types
import Foreign.C.String (castCharToCChar)

-- 2 fields

int32Double :: CInt -> Double ->  (CInt, Double)
int32Double = (,)

doubleInt32 :: Double -> CInt -> (Double, CInt)
doubleInt32 = (,)

charInt32 :: CChar -> CInt -> (CChar, CInt)
charInt32 = (,)

int32Char :: CInt -> CChar -> (CInt, CChar)
int32Char = (,)

-- Nested

list2A :: [CInt] -> String -> ([CInt], String)
list2A = (,)

list2B :: CInt -> [String] -> (CInt, [String])
list2B = (,)

-- 3 fields

int32DoubleChar :: CInt -> Double -> CChar ->  (CInt, Double, CChar)
int32DoubleChar = (,,)

doubleInt32Char :: Double -> CInt -> CChar -> (Double, CInt, CChar)
doubleInt32Char = (,,)

charInt32Double :: CChar -> CInt -> Double -> (CChar, CInt, Double)
charInt32Double = (,,)

int32CharChar :: CInt -> CChar -> CChar -> (CInt, CChar, CChar)
int32CharChar = (,,)

-- Nested

list3A :: [CInt] -> String -> CFloat -> ([CInt], String, CFloat)
list3A = (,,)

list3B :: CInt -> [String] -> CFloat -> (CInt, [String], CFloat)
list3B = (,,)

list3C :: CInt -> String -> [CFloat] -> (CInt, String, [CFloat])
list3C = (,,)

-- 4 fields

int32DoubleDoubleChar :: CInt -> Double -> Double -> CChar -> (CInt, Double, Double, CChar)
int32DoubleDoubleChar = (,,,)

doubleInt32CharChar :: Double -> CInt -> CChar -> CChar -> (Double, CInt, CChar, CChar)
doubleInt32CharChar = (,,,)

charInt32DoubleChar :: CChar -> CInt -> Double -> CChar -> (CChar, CInt, Double, CChar)
charInt32DoubleChar = (,,,)

int32CharCharChar :: CInt -> CChar -> CChar -> CChar -> (CInt, CChar, CChar, CChar)
int32CharCharChar = (,,,)

-- Nested

list4A :: [CInt] -> String -> CFloat -> Double -> ([CInt], String, CFloat, Double)
list4A = (,,,)

list4B :: CInt -> [String] -> CFloat -> Double -> (CInt, [String], CFloat, Double)
list4B = (,,,)

list4C :: CInt -> String -> [CFloat] -> Double -> (CInt, String, [CFloat], Double)
list4C = (,,,)

list4D :: CInt -> String -> CFloat -> [CDouble] -> (CInt, String, CFloat, [CDouble])
list4D = (,,,)

File: test/test_tuples.py
from hypothesis import given
from hypothesis.strategies import tuples, lists

from .context import pythas
from .t_types import *

types = pythas.types
utils = pythas.utils

import test.hs.testtuples as hst

# 2 fields
@given(c_ints, c_doubles)
def test_int32Double(a, b):
    assert hst.int32Double(a, b) == (a, b)


@given(c_doubles, c_ints)
def test_doubleInt32(a, b):
    assert hst.doubleInt32(a, b) == (a, b)


@given(c_chars, c_ints)
def test_charInt32(a, b):
    assert hst.charInt32(a, b) == (a, b)


@given(c_ints, c_chars)
def test_int32Char(a, b):
    assert hst.int32Char(a, b) == (a, b)


## Nested
@given(lists(c_ints), c_strings)
def test_list2A(a, b):
    assert hst.list2A(a, b) == (a, b)


@given(c_ints, lists(c_strings))
def test_list2B(a, b):
    assert hst.list2B(a, b) == (a, b)


# 3 fields
@given(c_ints, c_doubles, c_chars)
def test_int32DoubleChar(a, b, c):
    assert hst.int32DoubleChar(a, b, c) == (a, b, c)


@given(c_doubles, c_ints, c_chars)
def test_doubleInt32Char(a, b, c):
    assert hst.doubleInt32Char(a, b, c) == (a, b, c)


@given(c_chars, c_ints, c_doubles)
def test_charInt32Double(a, b, c):
    assert hst.charInt32Double(a, b, c) == (a, b, c)


@given(c_ints, c_chars, c_chars)
def test_int32CharChar(a, b, c):
    assert hst.int32CharChar(a, b, c) == (a, b, c)


## Nested
@given(lists(c_ints), c_strings, c_floats)
def test_list3A(a, b, c):
    assert hst.list3A(a, b, c) == (a, b, c)


@given(c_ints, lists(c_strings), c_floats)
def test_list3B(a, b, c):
    assert hst.list3B(a, b, c) == (a, b, c)


@given(c_ints, c_strings, lists(c_floats))
def test_list3C(a, b, c):
    assert hst.list3C(a, b, c) == (a, b, c)


# 4 fields
@given(c_ints, c_doubles, c_doubles, c_chars)
def test_int32DoubleDoubleChar(a, b, c, d):
    assert hst.int32DoubleDoubleChar(a, b, c, d) == (a, b, c, d)


@given(c_doubles, c_ints, c_chars, c_chars)
def test_doubleInt32CharChar(a, b, c, d):
    assert hst.doubleInt32CharChar(a, b, c, d) == (a, b, c, d)


@given(c_chars, c_ints, c_doubles, c_chars)
def test_charInt32DoubleChar(a, b, c, d):
    assert hst.charInt32DoubleChar(a, b, c, d) == (a, b, c, d)


@given(c_ints, c_chars, c_chars, c_chars)
def test_int32CharCharChar(a, b, c, d):
    assert hst.int32CharCharChar(a, b, c, d) == (a, b, c, d)


## Nested
@given(lists(c_ints), c_strings, c_floats, c_doubles)
def test_list4A(a, b, c, d):
    assert hst.list4A(a, b, c, d) == (a, b, c, d)


@given(c_ints, lists(c_strings), c_floats, c_doubles)
def test_list4B(a, b, c, d):
    assert hst.list4B(a, b, c, d) == (a, b, c, d)


@given(c_ints, c_strings, lists(c_floats), c_doubles)
def test_list4C(a, b, c, d):
    assert hst.list4C(a, b, c, d) == (a, b, c, d)


@given(c_ints, c_strings, c_floats, lists(c_doubles))
def test_list4C(a, b, c, d):
    assert hst.list4D(a, b, c, d) == (a, b, c, d)
